### ISSUE TITLE
How can I validate that SMTP is working correctly ? 

### ISSUE THREAD

**Issue by fgadot** [knadh/listmonk#288]
Hi
I am trying listmonk. 
I see a SMTP connection attempt on my mail server, but nothing after that. 
How can I debug what's going on on listmonk side, regarding SMTP ? 
Thank you

**Comment by knadh** [comment]
Have you checked Settings -> Logs? Any errors should be logged there.

**Comment by fgadot** [comment]
Yes. But there is not enough:

2021/03/09 08:48:39 loaded email (SMTP) messenger: admin@xxxx.com@xxx.com
2021/03/09 08:49:44 start processing campaign (test)
2021/03/09 08:49:44 campaign (test) finished

**Comment by knadh** [comment]
If there was an error when pushing an individual message, or with authentication, that would've been logged too. If there is nothing in the log, it means the SMTP didn't return any errors.

**Comment by fgadot** [comment]
I see that the SMTP Server does not return any error (from the SMTP itself). 
But once the connection is established, I do not see listmonk sending any data though (mail server in verbose debug), while it says that the test campaign is finished.

**Comment by fgadot** [comment]
So I found out more. 
While the logs say that the test campaign was finished, 3 minutes later I saw other logs this morning:

2021/03/09 09:36:15 manager.go:387: start processing campaign (test)
2021/03/09 09:36:15 manager.go:559: campaign (test) finished

2021/03/09 09:39:00 manager.go:283: error sending message in campaign test: subscriber xxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx: EOF
2021/03/09 09:39:00 manager.go:309: error sending message 'Finished: test': EOF

and on my email server
2021-03-09 09:39:00 TLS error on connection from [x.x.x.x]:51640 (SSL_accept): timed out
2021-03-09 09:39:00 TLS error on connection from [x.x.x.x]:51642 (SSL_accept): timed out

My email server requires TLS. I have this option enabled in the settings. I tried with and without TLS Verification without success. This email server manage over 100K Emails per hour without any issue from thousands of clients.

**Comment by knadh** [comment]
listmonk's getting an EOF (happens when the server terminates the connection abnormally), and the server is complaining of a TLS handshake time out, hmm. The app works fine with many different mailservers with TLS (SES, SendGrid, Postal, postfix etc.)

What SMTP server are you using?

**Comment by fgadot** [comment]
Dovecot as a MTA, with EXIM for the clients side.

**Comment by fgadot** [comment]
Cert is signed publicly. Supported encryption TLS1, TLS1.1, TLS1.1, SSLv3

**Comment by cgobbet** [comment]
I am experiencing a similar problem while sending the test message. 

Despite the right configuration, there is an EOF response too (with TLS activated or not). My mail service does have a working certificate too (I've just checked). I tried alternative SMTP configurations (like GMail), but the outcome is the same.

An odd detail: if instead of sending a test message, I run the campaign (with the same recipient, which is a GMail working address), the campaign goes through, everything seems to be ok, all green, but after some time, an EOF error message is shown again. ([screenshot](https://i.imgur.com/rR52q0o.png))

Recap:
- when trying to send a test message with all configurations in place → EOF error message
- when running the campaign to a test list with the same recipient → campaign ran showed as a success, but the message never reaches the recipient (and EOF error message too)

**Comment by knadh** [comment]
@cgobbet is your instance hosted on Heroku (or elsewhere)? If yes, can you describe the specifics of the environment. Since you have tried many SMTP servers and all of them produce EOF, it's got to be something in the environment. I'll try to recreate the same setup and try and debug. Thanks for your patience.

PS: There are many users who use listmonk on Heroku.

**Comment by cgobbet** [comment]
First, I am the one who thanks. You do an amazing job.

As I have lost considerable time here, I tried to document all my moves. Jump into TLDR if you find it too detailed.

Initially, I had LM installed on Heroku. I could not make it send the messages as I was having the EOF problem listed above. So, I installed Docker on my local machine to have more control of the environment. However, the problem remained the same (which leads me to believe that the problem is in the configuration, not the environment). As you can see on the screenshots, LM goes through the whole process, say that's ok (including the "finished" green status icons and an initial success log), but after a few minutes, it adds that the EOF error happened again (and, despite the error, the "finished" green icon remains).

Other than the steps to reproduce the problem, I'd like to add that the "Send test message" functionality does not work under any conditions (and the system correctly alerts that there was an EOF error).

Steps to reproduce the problem:

1) set up SMTP info
2) go to campaign
3) add recipient list and other details
4) click save and start campaign
5) process runs and after a few seconds, is marked as "finished"
6) log shows that Test campaign is finished
7) log adds two lines, saying that subscriber ID had an EOF error and Test campaign had an error.

TLDR:  SMTP setup does not work in different environments (Heroku and Docker for app). Campaign goes through, with all green icons, but after a few minutes, an error log is added.

Thanks a lot

![Campaign_run 4](https://user-images.githubusercontent.com/4549818/111748650-14785d00-8891-11eb-9f77-400ecc311603.png)
![Campaign_run 3](https://user-images.githubusercontent.com/4549818/111748656-15a98a00-8891-11eb-813e-9705a152dcfc.png)
![Campaign_run 2](https://user-images.githubusercontent.com/4549818/111748660-16dab700-8891-11eb-9208-a69ef9b49274.png)
![Campaign_run 1](https://user-images.githubusercontent.com/4549818/111748661-16dab700-8891-11eb-96d0-112a5700af28.png)

**Comment by knadh** [comment]
Looks like the exact issue @fgadot described. You can use the `Send test message` feature on the campaign page to quickly test sending mails. And you're right, this doesn't seem like an environment issue. It's either the config or some sort of an incompatibility between listmonk's SMTP libs and the upstream SMTP server.

That said, the only configuration issue that could cause this seems to be `TLS=on/off`. You have tried both with your SMTPs, right?

Since I'm unable to reproduce this in my local or any of our remote environments, this has to be debugged locally. @cgobbet would you be okay to get on a Meet/Zoom call and do a screenshare and try and debug this locally?

**Comment by cgobbet** [comment]
For the sake of consistency, I reinstalled the app from scratch (I had no data in there). The response was the same, with or without TLS (screenshot with setup added). I am available for a call, yes. Reach me by email, pls, and we can talk as soon as you need. Thanks!

![Screenshot 2021-03-19 at 11 35 14](https://user-images.githubusercontent.com/4549818/111767770-688e3c00-88a7-11eb-8db2-2c0a90ac69c4.png)

**Comment by manishiitg** [comment]
i am facing the same issue. not error comings in logs. but emails not going out

**Comment by manishiitg** [comment]
![image](https://user-images.githubusercontent.com/1370315/124859136-f2842100-dfcc-11eb-8ef2-a529f3087478.png)

**Comment by knadh** [comment]
`EOF` is the error. The SMTP server seems to be abruptly terminating connections. Is it possible for you to check server logs to see what's happening to connections coming from listmonk?

**Comment by acallaghan** [comment]
Hi there, I'm also seeing this issue as well. So it's a brand new installation, using the heroku build button.

I'm using Postmark as the SMTP server, and have not seen any emails getting delivered at all. When i've tailed the heroku logs, no errors are coming through for connecting the SMTP server. No inbound attempts are being seen on Postmark.

If I disable TLS, and try sending a test email, I get an error saying that the connection isn't secure. But if I re-enable it, it still doesn't work, and no errors are being seen.

The settings/logs section shows SMTP successes, and no errors as well

**Comment by acallaghan** [comment]
Perhaps this is happening on Heroku's hobby instances only? I'm not aware of any SMTP restrictions, but perhaps they silently block SMTP for the free instances to avoid spammy email bots?

**Comment by knadh** [comment]
> using the heroku build button.

Ah, unfortunately, the button (maintained by @bumi) doesn't deploy the latest version. Haven't been able to figure out a way to create a Heroku button that'll pick up the latest release and deploy instead of a hardcoded binary. @acallaghan any way for you to test out the latest version?

**Comment by acallaghan** [comment]
Ah okay, will do - I'll create a fork and try and deploy that directly & report back soon.

**Comment by bumi** [comment]
Yes, we need to update the deploy button to use the official versions directly. this is now possible as we no longer need nginx as a proxy. I started working on it but could not finish it. any help is welcome there. (jfyi. unrelated to this issue I guess)

**Comment by acallaghan** [comment]
Hey @bumi - so I tried this, I have a fork & updated the binary, it's still happening here. No SMTP is getting through at all from heroku, and no error logs. Just this:

```
2021-07-24T15:30:53.384692+00:00 heroku[router]: at=info method=POST path="/api/campaigns/1/test" host=***.herokuapp.com request_id=f252fb32-b543-42f7-8cc8-07bcbb244b6b fwd="89.238.143.237" dyno=web.1 connect=0ms service=61ms status=200 bytes=171 protocol=https
2021-07-24T15:30:53.385327+00:00 app[web.1]: measure#nginx.service=0.061 request_id=f252fb32-b543-42f7-8cc8-07bcbb244b6b
```

the only errors I can see ocurring is an api health / heartbeat type request?

```
2021-07-24T15:29:58.217068+00:00 heroku[router]: at=info method=GET path="/api/health" host=***.herokuapp.com request_id=8ca6c0cb-1f8c-44a3-98ae-24010be90a28 fwd="89.238.143.237" dyno=web.1 connect=0ms service=1ms status=502 bytes=295 protocol=https
2021-07-24T15:29:58.216636+00:00 app[web.1]: 2021/07/24 15:29:58 [error] 30#0: *118 connect() failed (111: Connection refused) while connecting to upstream, client: 10.39.143.244, server: _, request: "GET /api/health HTTP/1.1", upstream: "http://127.0.0.1:9000/api/health", host: "***.herokuapp.com", referrer: "https://***.herokuapp.com/settings"
2021-07-24T15:29:58.216699+00:00 app[web.1]: measure#nginx.service=0.000 request_id=8ca6c0cb-1f8c-44a3-98ae-24010be90a28
2021-07-24T15:29:58.280310+00:00 app[web.1]: 2021/07/24 15:29:58 init.go:439: media upload provider: filesystem
2021-07-24T15:29:58.716726+00:00 app[web.1]: 2021/07/24 15:29:58 [error] 30#0: *120 connect() failed (111: Connection refused) while connecting to upstream, client: 10.39.143.244, server: _, request: "GET /api/health HTTP/1.1", upstream: "http://127.0.0.1:9000/api/health", host: "***.herokuapp.com", referrer: "https://***.herokuapp.com/settings"
2021-07-24T15:29:58.716757+00:00 app[web.1]: measure#nginx.service=0.000 request_id=fa7014b8-23b8-4fe7-a31e-12d9f3a1853c
```

Any ideas? If you'd like, message me privately and i can make you a collaborator of my project

**Comment by acallaghan** [comment]
I tested this using a hobby dyno instead of a free one, and still no SMTP getting through - any help/ideas as to why this might not be working?

**Comment by bumi** [comment]
Heroku should not be a problem and also should not require special settings.
could you confirm the settings with another server? does it work there?
sorry, I don’t know much here.

**Comment by Citrullin** [comment]
I had some issues with the SMTP connection before. For me the issue was quite simple: I forgot to enable the SMTP port on the other vServer. Both server mailserver and app-server need this port to be not restricted. For security reasons some cloud provider disable it by default.

**Comment by knadh** [comment]
Closing this issue as it's inconclusive. Nothing so far that indicates that there may be an issue with how listmonk connects to SMTP servers. `EOF` just indicates abruptly terminated TCP connections. Thanks for your patience!

**Comment by fwarren** [comment]
Since you can not really debug or troubleshoot the GO module that does SMTP I have found running my own email relay has been useful. 

Add to the `docker-compose.yml` file the following:
```
  postfix:
    image: juanluisbaptiste/postfix
    restart: always
    environment:
      - SMTP_SERVER=smtp.my-mail-provider.com
      - SMTP_USERNAME=webmaster@mywebsite.com
      - SMTP_PASSWORD=really-good-password
      - SERVER_HOSTNAME=mywbsite.com
      # delete this line or find out what network your containers are on
      - SMTP_NETWORKS=127.0.0.0/8,172.22.0.0/16
```
In Listmonk Settings -> Settings -> SMTP
* Host: postfix
* Port: 25
* Auth Protocol: none

At which point I can monitor what is being sent :
```
docker-compose logs -f postfix
```

**Comment by sumitmamoria** [comment]
I have been trying the latest version of listmonk. I have a custom SMTP server which is known to be working for a long time. When I send test mail from the SMTP tab, it throws an error - "dial tcp x.x.x.x:587: i/o timeout". Any ideas ?

**Comment by knadh** [comment]
Seems like a Firewall issue. Could be at the network level or could be something like SELinux.

**Comment by Citrullin** [comment]
> I have been trying the latest version of listmonk. I have a custom SMTP server which is known to be working for a long time. When I send test mail from the SMTP tab, it throws an error - "dial tcp x.x.x.x:587: i/o timeout". Any ideas ?

Check the logs and trace everything. It was just a wrong config in my case. It's most likely to some issue like that. Just dig deep enough and understand where is happening what. You will find the issue :)
